src/interface/blas2_interface.hpp

```diff
@@ -733,16 +733,73 @@ typename sb_handle_t::event_t _spr_impl(sb_handle_t& sb_handle, char _Uplo,
   const index_t globalSize = localSize * nWGPerCol;
 
   if (Upper) {
-    auto spr = make_gerp<true, true>(mA, _N, _alpha, vx, _incx, vx, _incx);
+    auto spr = make_spr<true, true>(mA, _N, _alpha, vx, _incx, vx, _incx);
     return ret = concatenate_vectors(
                ret, sb_handle.execute(spr, localSize, globalSize));
   } else {
-    auto spr = make_gerp<true, false>(mA, _N, _alpha, vx, _incx, vx, _incx);
+    auto spr = make_spr<true, false>(mA, _N, _alpha, vx, _incx, vx, _incx);
     return ret = concatenate_vectors(
                ret, sb_handle.execute(spr, localSize, globalSize));
   }
 }
 
+/*! _SPR2.
+ * @brief Implementation of the rank 2 operation
+ */
+/*
+sspr2 	( 	character  	UPLO,
+   integer  	N,
+   real  	ALPHA,
+   real, dimension(N)  	X,
+   integer  	INCX,
+   real, dimension(N)  	Y,
+   integer  	INCY,
+   real, dimension(N, N + 1 / 2)  	AP
+ )
+*/
+template <typename sb_handle_t, typename index_t, typename element_t,
+          typename container_t0, typename increment_t, typename container_t1,
+          typename container_t2>
+typename sb_handle_t::event_t _spr2_impl(sb_handle_t& sb_handle, char _Uplo,
+                                         index_t _N, element_t _alpha,
+                                         container_t0 _vx, increment_t _incx,
+                                         container_t1 _vy, increment_t _incy,
+                                         container_t2 _mPA) {
+  // throw exception if invalid arguments
+  if (_N <= 0) {
+    throw std::invalid_argument("Invalid vector size");
+  }
+
+  // bail out early if alpha == 0
+  if (_alpha == (element_t)0) {
+    std::vector<cl::sycl::event> event;
```

Comment:
Maybe change this from `std::vector<>` to `typename sb_handle_t::event_t`? Same for line `719` as well.